In [4]:
# importing the data in csv file

DiabetesDataset = pd.read_csv('/content/diabetes.csv')
DiabetesDataset.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [5]:
import pandas as pd # for manipulating and loading the dataset
import numpy as np # for creating arrays of the dataset
from sklearn.preprocessing import StandardScaler # it use to bring the data in specific range which is imp step before training of model
from sklearn import svm
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In The above dataset column i will say outcome is the label if 1 = then patient is diabetic
if 0 = then patient is not diabetic

In [6]:
DiabetesDataset.shape # 768 data of womens and 9  attributes

(768, 9)

In [7]:
DiabetesDataset.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [8]:
# counting number of ones and number of zeroes
DiabetesDataset['Outcome'].value_counts() # groups the data say 1 group of zeroes and other group of ones and then count their presence

0    500
1    268
Name: Outcome, dtype: int64

In [9]:
DiabetesDataset.groupby('Outcome').mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

when glucose level low 109 around no diabetes
when glucose level high 141 around then diabetes .. this is somewhat what ml algo will do

In [10]:
# separating the data and labels

X = DiabetesDataset.drop(columns = 'Outcome',axis = 1)
Y = DiabetesDataset['Outcome']

In [11]:
X.head() # label column i.e i will say outcome is removed

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  
0                     0.627   50  
1                     0.351   31  
2                     0.672   32  
3                     0.167   21  
4                     2.288   33

In [12]:
X.shape

(768, 8)

In [13]:
Y.head()

0    1
1    0
2    1
3    0
4    1
Name: Outcome, dtype: int64

In [14]:
# Data Standardization

# As  data lies in different ranges it becomes difficult for model training so we have to normalize the data.
# Normalizing the data means bringing the data in same range
scaler = StandardScaler() # created instance of the standard scaler class

scaler.fit(X)

standardizedData = scaler.transform(X) # can also use scaler.fit_transform()

Scaling

After calling scaler.fit(data) on a 2D array data, the scaler object will compute the mean and standard deviation for each feature (column) in the input data.
 The scaler object will store these computed parameters internally, and they will be used later when you apply the scaling transformation using scaler.
 transform(data).
 Specifies parameter using which tranformation will takes place .

 Transformation

 The transform method is used to apply the scaling or standardization transformation to the data based on the parameters learned during the fitting process.
It takes the data as input and returns the transformed data where the scaling or standardization has been applied according to the learned parameters.

In [15]:
print(standardizedData) # all values are in similar range  -- row wise array banata hai

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


In [16]:
standardizedData.shape

(768, 8)

In [17]:
X=standardizedData  # X contains standardized dataset without Labels
Y=DiabetesDataset['Outcome'] # Y contains Labels

In [18]:
X

array([[ 0.63994726,  0.84832379,  0.14964075, ...,  0.20401277,
         0.46849198,  1.4259954 ],
       [-0.84488505, -1.12339636, -0.16054575, ..., -0.68442195,
        -0.36506078, -0.19067191],
       [ 1.23388019,  1.94372388, -0.26394125, ..., -1.10325546,
         0.60439732, -0.10558415],
       ...,
       [ 0.3429808 ,  0.00330087,  0.14964075, ..., -0.73518964,
        -0.68519336, -0.27575966],
       [-0.84488505,  0.1597866 , -0.47073225, ..., -0.24020459,
        -0.37110101,  1.17073215],
       [-0.84488505, -0.8730192 ,  0.04624525, ..., -0.20212881,
        -0.47378505, -0.87137393]])

In [19]:
Y

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64

In [20]:
# Split the dataset into training and testing dataset

Xtrain , Xtest , Ytrain , Ytest = train_test_split(X,Y,test_size=0.2,stratify=Y,random_state=2)

In [21]:
print(Xtrain)

[[-1.14185152 -0.05929342 -3.57259724 ...  0.05170968 -0.9992857
  -0.78628618]
 [ 0.63994726 -0.49745345  0.04624525 ... -0.15136112 -1.05666795
   0.31985461]
 [-0.84488505  2.13150675 -0.47073225 ... -0.24020459 -0.2231152
   2.19178518]
 ...
 [ 2.12477957 -1.12339636  0.25303625 ... -0.24020459 -0.51908683
   0.14967911]
 [ 0.04601433 -0.27837344  0.45982725 ...  0.94014439 -0.71237443
   0.40494237]
 [-1.14185152 -1.09209922 -0.05715025 ...  0.48323511 -0.70633419
  -0.70119842]]


In [22]:
Xtrain.shape

(614, 8)

In [23]:
print(Ytrain)

619    1
329    0
13     1
476    1
45     1
      ..
303    1
592    1
559    0
725    0
253    0
Name: Outcome, Length: 614, dtype: int64


In [24]:
Ytrain.shape

(614,)

In [25]:
print(Xtest)

[[-0.25095213 -0.46615631  0.14964075 ... -0.78595734 -0.79995787
  -0.53102292]
 [-0.25095213 -0.24707629 -1.29789624 ... -1.21748278 -1.00230582
  -0.95646168]
 [ 0.3429808   0.81702665  0.45982725 ...  0.2167047  -0.76673656
   2.7023117 ]
 ...
 [-0.54791859 -0.34096773  0.25303625 ...  0.05170968  0.6829204
  -0.53102292]
 [-0.54791859 -1.24858494  0.14964075 ... -0.24020459  0.22688248
  -0.70119842]
 [ 2.7187125  -0.46615631  0.14964075 ...  0.58477051 -0.88754131
   1.00055664]]


In [26]:
Xtest.shape

(154, 8)

In [27]:
print(Ytest)

615    0
80     0
148    0
132    1
501    0
      ..
127    0
185    1
85     0
497    0
86     0
Name: Outcome, Length: 154, dtype: int64


In [28]:
Ytest.shape

(154,)

In [29]:
classifier = svm.SVC(kernel='linear')


In [30]:
# training svm classifier on the trAIning dataset
diabetesModel = classifier.fit(Xtrain,Ytrain)

In [31]:
# lets look at accuracy of trained model on training data set

trainedDataPrediction = classifier.predict(Xtrain) # gives the outcome as output

accuracyOnTrainedData = accuracy_score(trainedDataPrediction,Ytrain) # checking predicted outcome and actual outcome

In [32]:
print(trainedDataPrediction)

[0 0 1 0 1 1 0 1 0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0
 0 0 0 0 1 0 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0
 1 0 0 0 0 0 0 1 1 0 0 0 0 1 0 0 0 0 1 0 0 1 0 0 1 1 0 1 1 0 0 0 0 0 0 0 1
 0 0 0 0 0 0 0 1 0 1 0 1 0 0 1 1 1 0 1 0 1 1 0 0 0 0 1 0 0 0 0 0 0 1 0 0 0
 0 1 0 1 1 0 1 0 0 1 1 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 1 0 1 0 0 0 0 1 0 0 1
 0 0 0 1 0 0 0 1 0 0 1 0 0 1 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0
 0 1 0 0 0 0 0 0 0 1 0 0 0 1 1 0 1 1 0 1 0 0 0 0 0 0 1 0 0 1 0 0 0 0 1 1 1
 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 0
 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 1 1 0 1 0 0 0 0 0 0 0 1 0 0 0 1
 0 0 0 1 0 0 0 0 1 0 0 1 0 0 1 1 0 1 0 0 0 1 0 0 0 0 1 0 0 0 1 1 0 1 0 0 0
 0 1 1 1 0 0 1 1 1 0 1 0 0 0 1 0 1 0 0 0 1 0 0 1 1 0 0 0 0 1 1 0 0 1 0 0 0
 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0 1 1 0 0 0 0 1 0 0 0 1
 0 1 0 0 1 1 1 0 0 0 1 0 0 0 1 1 0 0 0 0 0 1 1 0 0 0 0 1 1 0 0 0 0 0 1 0 0
 0 0 0 0 1 0 1 0 0 0 1 0 

In [33]:
# lets check accuracy for the test dataset

testAccuracyData = diabetesModel.predict(Xtest)

testDataAccuracy = accuracy_score(testAccuracyData,Ytest)

print('test data accuracy = ' , testDataAccuracy)

test data accuracy =  0.7727272727272727


In [34]:
# lets take input from user and try to predict


data = (4,110,92,0,0,37.6,0.191,30)

data = (8,125,96,0,0,0,0.232,54)



# model is trained to handle 738 or something amount on rows so if directly give model this single row data it will get confuse so have to reshape it
# and also have to bring the data into same range i.e to same data manipulation that we while creating model

# first of lets convert list data into numpy array as fast processing , easy manipulations

input_array_data = np.asarray(data)

# reshaping the data
reshapeInputData = input_array_data.reshape(1,-1)

# (1, -1) specify that you want to reshape it into a 2-dimensional array with one row and as many columns as needed to maintain the total number of elements.

print(reshapeInputData)


[[  8.    125.     96.      0.      0.      0.      0.232  54.   ]]


In [35]:
# lets apply same standardization that we had applied previously

transInputData = scaler.transform(reshapeInputData)

print(transInputData)

[[ 1.23388019  0.12848945  1.39038675 -1.28821221 -0.69289057 -4.06047387
  -0.7244549   1.76634642]]


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


In [36]:
# lets predict the outcome

output = diabetesModel.predict(transInputData)

if output==1:
  print('patient is infected with diabetes')
else:
  print('patient is not infected with diabetes')



patient is not infected with diabetes


In [37]:
# saving the model that we trained

# as we have perform standardization on the dataset wnd created model we have to save model as well as standardized
# parameter

import pickle

with open('diabetesModel.pkl', 'wb') as model_file:
    pickle.dump(diabetesModel, model_file)

# Save StandardScaler
with open('standard_scaler.pkl', 'wb') as scaler_file:
    pickle.dump(scaler, scaler_file)

In [38]:
# Loading the model and trying to manipulate (means checking whether we can load this model anywhere and still it is going
# output  )

# Load SVM model
with open('diabetesModel.pkl', 'rb') as model_file:
    loaded_svm_model = pickle.load(model_file)

# Load StandardScaler
with open('standard_scaler.pkl', 'rb') as scaler_file:
    loaded_scaler = pickle.load(scaler_file)




In [39]:
data = (4,110,92,0,0,37.6,0.191,30)

data = (8,125,96,0,0,0,0.232,54)

data = (5,166,72,19,175,25.8,0.587,51)

input_array_data = np.asarray(data)

# reshaping the data
reshapeInputData = input_array_data.reshape(1,-1)

# (1, -1) specify that you want to reshape it into a 2-dimensional array with one row and as many columns as needed to maintain the total number of elements.


scaleDta =loaded_scaler.transform(reshapeInputData) # using loaded model

output = loaded_svm_model.predict(scaleDta) # using loaded model

print(output)



[1]


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


In [40]:
from google.colab import files

# Download the model file
files.download('diabetesModel.pkl') # code for downloadinng the  model file o local machine

# Download the scaler file
files.download('standard_scaler.pkl')# code for downloadinng the  model file o local machine


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>